File: two_pointers/triplet_sum_close_to_target/my_solution.java
import java.util.*;

class TripletSumCloseToTarget {

  public static int searchTriplet(int[] arr, int targetSum) {
    // TODO: Write your code here
    Arrays.sort(arr);
    int minTripletSumCloseToTarget = Integer.MAX_VALUE;
    int minDisplacementFromSum = Integer.MAX_VALUE;
    System.out.println("Target: " + targetSum + "\t\t" + Arrays.toString(arr));
    for (int i = 0; i < arr.length; i++) {
        if ( ( i > 0 ) && ( arr[i] == arr[i-1] ) )
            continue;

        int currentTripletSum =  arr[i] + closestTwoSum(arr, targetSum - arr[i], i);
        int currentDisplacementFromSum = Math.abs( targetSum - currentTripletSum  );
        if ( currentDisplacementFromSum < minDisplacementFromSum ) {
            minTripletSumCloseToTarget = currentTripletSum;
            minDisplacementFromSum = currentDisplacementFromSum;
        } else if ( currentDisplacementFromSum == minDisplacementFromSum ) {
            minTripletSumCloseToTarget = Math.min( 
                minTripletSumCloseToTarget, 
                currentTripletSum
            );
        }
        System.out.print("currentTripletSum: " + currentTripletSum);
        System.out.print(" currentDisplacement: " + currentDisplacementFromSum);
        System.out.println(" minTripletSum: " + minTripletSumCloseToTarget);
    }
    return minTripletSumCloseToTarget;
  }

  public static int closestTwoSum(int[] arr, int targetSum, int index) {
      int left = 0;
      int right = arr.length - 1;
      int minDisplacementFromSum = Integer.MAX_VALUE;
      int closestTwoSum = Integer.MAX_VALUE;

      while ( left < right ) {
          if ( left == index )
              left += 1;
          if ( right == index )
              right -= 1;

          System.out.print("\t\tTwoSumTarget: " + targetSum + " Left: " + left + " Right: " + right);
          System.out.println(" minDisplacement: " + minDisplacementFromSum + " minTwoSum: " + closestTwoSum);

          int currentSum = arr[right] + arr[left];
          int currentDisplacementFromSum = Math.abs( targetSum - ( arr[right] + arr[left] ) );
          if ( currentDisplacementFromSum < minDisplacementFromSum ) {
              closestTwoSum = currentSum;
              minDisplacementFromSum = currentDisplacementFromSum;
          }

          if ( currentSum > targetSum )
              right--;
          else
              left++;

          System.out.print("\t\tTwoSumTarget: " + targetSum + " Left: " + left + " Right: " + right);
          System.out.println(" minDisplacement: " + minDisplacementFromSum + " minTwoSum: " + closestTwoSum);
      }

      System.out.println("\t\tResult: \t\tTarget: " + targetSum + " currentMinTwoSum: " + closestTwoSum);

      return closestTwoSum;
  }
}

File: two_pointers/quadruplet_sum_to_target/my_solution.js
const search_quadruplets = function(arr, target) {
  quadruplets = [];
  // TODO: Write your code here
  arr.sort( (a, b) => a - b );

  console.log(arr);
  console.log(target);

  for ( let i = 0; i < arr.length - 3; i++ ) {
      for ( let j = i + 1; j < arr.length - 2; j++) {
          quadruplets = two_sum(arr, j+1, target, arr[i], arr[j], quadruplets);
      }
  }
  return quadruplets;
};ii

const two_sum = function(arr, start, target, firstNum, secondNum, solutions) {
  var right = arr.length - 1;
  var left = start;

  console.log( "firstNum : " + firstNum  + " secondNum : " + secondNum );

  while ( left < right ) {
      console.log( "\t\tLeft - " + left + " : " + arr[left]);
      console.log( "\t\tRight - " + right + " : " + arr[right]);

      var current_sum = arr[left] + arr[right] + firstNum + secondNum;
      console.log("\t\tSum: " + current_sum);
      
      if ( current_sum === target ) {
          
          potential_array = [ firstNum, secondNum, arr[left], arr[right] ];
          existsAt = solutions.findIndex( function(elem, index) {
            return equals(elem, potential_array);
          });
          if ( existsAt === -1 ){  
              solutions.push( [ firstNum, secondNum, arr[left], arr[right] ] );
          }
          left += 1;

      } 
      else if ( current_sum < target ) { left += 1; } 
      else { right -= 1; }

      console.log( "\t\tLeft - " + left + " : " + arr[left]);
      console.log( "\t\tRight - " + right + " : " + arr[right]);

  }

  return solutions;
}

const equals = (a, b) => { 
  return JSON.stringify(a) === JSON.stringify(b); 
}
